// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: testdata.proto

package testdata

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/types/known/anypb"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = anypb.Any{}
	_ = sort.Sort
)

// Validate checks the field values on TestRecord with the rules defined in the
// proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *TestRecord) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on TestRecord with the rules defined in
// the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in TestRecordMultiError, or
// nil if none found.
func (m *TestRecord) ValidateAll() error {
	return m.validate(true)
}

func (m *TestRecord) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Double_

	// no validation rules for Float_

	// no validation rules for Int32_

	// no validation rules for Int64_

	// no validation rules for Uint32_

	// no validation rules for Uint64_

	// no validation rules for Sint32_

	// no validation rules for Sint64_

	// no validation rules for Fixed32_

	// no validation rules for Fixed64_

	// no validation rules for Sfixed32_

	// no validation rules for Sfixed64_

	// no validation rules for Bool_

	// no validation rules for String_

	// no validation rules for DoubleBase_

	// no validation rules for FloatBase_

	// no validation rules for Int32Base_

	// no validation rules for Int64Base_

	// no validation rules for Uint32Base_

	// no validation rules for Uint64Base_

	// no validation rules for Sint32Base_

	// no validation rules for Sint64Base_

	// no validation rules for Fixed32Base_

	// no validation rules for Fixed64Base_

	// no validation rules for Sfixed32Base_

	// no validation rules for Sfixed64Base_

	// no validation rules for BoolBase_

	// no validation rules for StringBase_

	// no validation rules for TestRecordId

	// no validation rules for DeletedAt

	// no validation rules for CreatedAt

	// no validation rules for UpdatedAt

	if len(errors) > 0 {
		return TestRecordMultiError(errors)
	}

	return nil
}

// TestRecordMultiError is an error wrapping multiple validation errors
// returned by TestRecord.ValidateAll() if the designated constraints aren't met.
type TestRecordMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m TestRecordMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m TestRecordMultiError) AllErrors() []error { return m }

// TestRecordValidationError is the validation error returned by
// TestRecord.Validate if the designated constraints aren't met.
type TestRecordValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e TestRecordValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e TestRecordValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e TestRecordValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e TestRecordValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e TestRecordValidationError) ErrorName() string { return "TestRecordValidationError" }

// Error satisfies the builtin error interface
func (e TestRecordValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sTestRecord.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = TestRecordValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = TestRecordValidationError{}

// Validate checks the field values on DatetimeRecord with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *DatetimeRecord) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on DatetimeRecord with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in DatetimeRecordMultiError,
// or nil if none found.
func (m *DatetimeRecord) ValidateAll() error {
	return m.validate(true)
}

func (m *DatetimeRecord) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetDatetime1_()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, DatetimeRecordValidationError{
					field:  "Datetime1_",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, DatetimeRecordValidationError{
					field:  "Datetime1_",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetDatetime1_()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return DatetimeRecordValidationError{
				field:  "Datetime1_",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if all {
		switch v := interface{}(m.GetDatetime3_()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, DatetimeRecordValidationError{
					field:  "Datetime3_",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, DatetimeRecordValidationError{
					field:  "Datetime3_",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetDatetime3_()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return DatetimeRecordValidationError{
				field:  "Datetime3_",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if all {
		switch v := interface{}(m.GetDatetime5_()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, DatetimeRecordValidationError{
					field:  "Datetime5_",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, DatetimeRecordValidationError{
					field:  "Datetime5_",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetDatetime5_()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return DatetimeRecordValidationError{
				field:  "Datetime5_",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if all {
		switch v := interface{}(m.GetDatetime7_()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, DatetimeRecordValidationError{
					field:  "Datetime7_",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, DatetimeRecordValidationError{
					field:  "Datetime7_",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetDatetime7_()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return DatetimeRecordValidationError{
				field:  "Datetime7_",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if all {
		switch v := interface{}(m.GetDatetime9_()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, DatetimeRecordValidationError{
					field:  "Datetime9_",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, DatetimeRecordValidationError{
					field:  "Datetime9_",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetDatetime9_()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return DatetimeRecordValidationError{
				field:  "Datetime9_",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return DatetimeRecordMultiError(errors)
	}

	return nil
}

// DatetimeRecordMultiError is an error wrapping multiple validation errors
// returned by DatetimeRecord.ValidateAll() if the designated constraints
// aren't met.
type DatetimeRecordMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m DatetimeRecordMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m DatetimeRecordMultiError) AllErrors() []error { return m }

// DatetimeRecordValidationError is the validation error returned by
// DatetimeRecord.Validate if the designated constraints aren't met.
type DatetimeRecordValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e DatetimeRecordValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e DatetimeRecordValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e DatetimeRecordValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e DatetimeRecordValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e DatetimeRecordValidationError) ErrorName() string { return "DatetimeRecordValidationError" }

// Error satisfies the builtin error interface
func (e DatetimeRecordValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sDatetimeRecord.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = DatetimeRecordValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = DatetimeRecordValidationError{}
